src/main/java/com/linkedin/kmf/services/ProduceService.java

```diff
@@ -257,16 +288,18 @@ void addPartitionSensors(int partition) {
    */
   private class ProduceRunnable implements Runnable {
     private final int _partition;
+    private final String _key;
 
-    ProduceRunnable(int partition) {
+    ProduceRunnable(int partition, String key) {
       _partition = partition;
+      _key = key;
     }
 
     public void run() {
       try {
-        long nextIndex = _nextIndexPerPartition.get(_partition).get();
+        long nextIndex = _nextIndexPerKey.get(_key).get();
         String message = Utils.jsonFromFields(_topic, nextIndex, System.currentTimeMillis(), _producerId, _recordSize);
-        BaseProducerRecord record = new BaseProducerRecord(_topic, _partition, null, message);
+        BaseProducerRecord record = new BaseProducerRecord(_topic, null, _key, message);
```

Comment:
But that is based on the assumption that the underlying `KMBaseProducer` will use the same hashing algorithm to map key to the partition, right? I am not sure every implementation of `KMBaseProducer` will do that. For example, what if `KafkaRestProducer` will be extended in the future to allow user to specify the destination partition but not the message key?